third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,512 @@
+        let to_label = self.get_block_label(to_block).expect("label");
+        let attr_id = self
+            .code_blocks
+            .get(&to_block)
+            .expect("code block")
+            .first()
+            .expect("first instruction")
+            .get_attr_id();
+        codes.push(Bytecode::Jump(attr_id, to_label));
+    }
+
+    /// Returns the instructions of the block
+    fn block_instrs(&self, block_id: BlockId) -> &[Bytecode] {
+        self.code_blocks.get(&block_id).expect("block instructions")
+    }
+
+    /// Returns the label of the block or `None` if it doesn't start with a label
+    fn get_block_label(&self, block_id: BlockId) -> Option<Label> {
+        if let Bytecode::Label(_, label) = self
+            .block_instrs(block_id)
+            .first()
+            .expect("first instruction")
+        {
+            Some(*label)
+        } else {
+            None
+        }
+    }
+
+    /// Checks if the block is entry or exit block
+    fn is_trivial_block(&self, block: BlockId) -> bool {
+        block == self.cfg.entry_block() || block == self.cfg.exit_block()
+    }
+}
+
+struct RemoveEmptyBlock(ControlFlowGraphCodeGenerator);
+
+/// `impl_deref!(Wrapper, WrappedType);` implements `Deref` trait for a wrapper struct with one field
+/// `struct Wrapper(WrappedType);` where the `deref` method simply returns a reference to the wrapped value.
+macro_rules! impl_deref {
+    ($struct_name:ident, $target_type:ty) => {
+        impl std::ops::Deref for $struct_name {
+            type Target = $target_type;
+
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+    };
+}
+
+/// `impl_deref!(Wrapper)` implements `DerefMut` trait for a wrapper struct with one field,
+/// where the `deref_mut` method simply returns a mutable reference to the wrapped value.
+/// Note that `Wrapper` should've implemented `Deref`, which is a supertrait of `DerefMut`.
+macro_rules! impl_deref_mut {
+    ($struct_name:ident) => {
+        impl std::ops::DerefMut for $struct_name {
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                &mut self.0
+            }
+        }
+    };
+}
+
+impl_deref!(RemoveEmptyBlock, ControlFlowGraphCodeGenerator);
+
+impl_deref_mut!(RemoveEmptyBlock);
+
+impl RemoveEmptyBlock {
+    /// Wrapper
+    pub fn new(generator: ControlFlowGraphCodeGenerator) -> Self {
+        Self(generator)
+    }
+
+    /// - Eliminates branch/jump to jump
+    ///     L1: goto L2 (L1 != L2)
+    ///     =>
+    ///     (removed)
+    ///
+    ///     goto L1
+    ///     =>
+    ///     goto L2
+    ///
+    ///     if ... goto L1 else ... / if ... goto ... else L1
+    ///     =>
+    ///     if ... goto L2 else ... / if ... goto ... else L2
+    pub fn transform(&mut self) {
+        for block in self.cfg.blocks() {
+            if self.is_empty_block(block) {
+                let suc_block = self.get_the_non_trivial_successor(block);
```

Comment:
Add a comment that the successor is guaranteed to be unique because the empty block ends in a Jump

Reply:
?

Reply:
outdated